### PR TITLE
feat: Public API for listing workflow runs

### PR DESCRIPTION
**Pull request by jamesclark-Zapata**
# The problem

There is no way to get the list of workflow runs via the public API

# This PR's solution

- Adds a new function to the public API to list workflow runs by config
- Adds a new method to the internal runtime interface

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/internal/docs>[!]:`

### CODE REVIEW

File: src/orquestra/sdk/_base/_api.py [hunk 1]
```diff
@@ -663,6 +666,124 @@ def get_tasks(self) -> t.Set[TaskRun]:
...
+        "For example:\n"
+        '- "9h32m" = 9 hours and 32 minutes,\n'
+        '- "8H6S" = 8 hours and 6 seconds,\n'
+        '- "10m" = 10 minutes,\n'
+        '- "3D6h8M13s" = 3 days, 6 hours, 8 minutes and 13 seconds.'
+    )
+
+
+def list_workflow_runs(
+    config_name: str,
```
Comment (jamesclark-Zapata) [inline src/orquestra/sdk/_base/_api.py:727]: Use config objects in addtion to `ConfigName`
Also switch `str` to `ConfigName`.

File: src/orquestra/sdk/_base/_api.py [hunk 2]
```diff
@@ -663,6 +666,124 @@ def get_tasks(self) -> t.Set[TaskRun]:
...
+    *,
+    limit: t.Optional[int] = None,
+    prefix: t.Optional[str] = None,
+    max_age: t.Optional[str] = None,
+    state: t.Optional[State] = None,
```
Comment (jamesclark-Zapata) [inline src/orquestra/sdk/_base/_api.py:732]: Allow list of States

File: src/orquestra/sdk/_base/_api.py [hunk 3]
```diff
@@ -663,6 +666,124 @@ def get_tasks(self) -> t.Set[TaskRun]:
...
+        max_age: Only return runs younger than the specified maximum age.
+        status: Only return runs of runs with the specified status.
+        project_dir: The location of the project directory. This directory must
+            contain the workflows database to which this run was saved. If omitted,
+            the current working directory is assumed to be the project directory.
+        config_save_file: The location to which the associated configuration was
+            saved. If omitted, the default config file path is used.
+
+    Raises:
+        ModuleNotFoundError: when the orquestra-runtime package is not installed.
```
Comment (jamesclark-Zapata) [inline src/orquestra/sdk/_base/_api.py:751]: Remove this

File: src/orquestra/sdk/_base/_api.py [hunk 4]
```diff
@@ -663,6 +666,124 @@ def get_tasks(self) -> t.Set[TaskRun]:
...
+        '- "10m" = 10 minutes,\n'
+        '- "3D6h8M13s" = 3 days, 6 hours, 8 minutes and 13 seconds.'
+    )
+
+
+def list_workflow_runs(
+    config_name: str,
+    *,
+    limit: t.Optional[int] = None,
+    prefix: t.Optional[str] = None,
```
Comment (jamesclark-Zapata) [inline src/orquestra/sdk/_base/_api.py:730]: Remove prefix

File: src/orquestra/sdk/_base/_driver/_ce_runtime.py [hunk 5]
```diff
@@ -249,7 +255,32 @@ def stop_workflow_run(self, workflow_run_id: WorkflowRunId) -> None:
...
+        Args:
+            limit: Restrict the number of runs to return, prioritising the most recent.
+            prefix: Only return runs that start with the specified string.
+            max_age: Only return runs younger than the specified maximum age.
+            status: Only return runs of runs with the specified status.
+        Returns:
+                A list of the workflow runs
+        """
+        try:
+            runs = self._client.list_workflow_runs()
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/_driver/_ce_runtime.py:280]: **suggestion**: as of today the filters aren't passed to the `client`. I suppose we will pass them in the future. Is there a ticket we can mention in a comment here?

File: src/orquestra/sdk/_base/_qe/_qe_runtime.py [hunk 6]
```diff
@@ -816,3 +816,55 @@ def stop_workflow_run(self, run_id: WorkflowRunId) -> None:
...
             ) from e
+
+    def list_workflow_runs(
+        self,
+        *,
+        limit: Optional[int] = None,
+        prefix: Optional[str] = None,
+        max_age: Optional[timedelta] = None,
+        state: Optional[Union[State, List[State]]] = None,
+    ) -> List[WorkflowRun]:
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/_qe/_qe_runtime.py:827]: **suggestion**: there are errors we expect this method to raise, like `UnauthorizedError`. Could you list them in the docstring? This helps with code and test isolation.

File: src/orquestra/sdk/_base/abc.py [hunk 7]
```diff
@@ -179,6 +185,28 @@ def iter_logs(
         """
         raise NotImplementedError()
 
+    @abstractmethod
+    def list_workflow_runs(
+        self,
+        *,
+        limit: t.Optional[int] = None,
+        prefix: t.Optional[str] = None,
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/abc.py:193]: **question**: you mentioned earlier that we planned to not use the ID prefix. We don't use it in the Python API. Was it intentional to put this param here?

**Review by alexjuda — APPROVED** [review]
Added few non-blocking comments

File: src/orquestra/sdk/_base/abc.py (hunk 7 above)
Comment (jamesclark-Zapata) [reply src/orquestra/sdk/_base/abc.py]: done in [a2d40d9](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/3/commits/a2d40d91e2148521c57fc627af37fc24c462d5bd)

File: src/orquestra/sdk/_base/_qe/_qe_runtime.py (hunk 6 above)
Comment (jamesclark-Zapata) [reply src/orquestra/sdk/_base/_qe/_qe_runtime.py:826]: done in [f4b0dc9](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/3/commits/f4b0dc9ebbf762c0ef05d129f8b38db383a12c00)

File: src/orquestra/sdk/_base/_driver/_ce_runtime.py (hunk 5 above)
Comment (jamesclark-Zapata) [reply src/orquestra/sdk/_base/_driver/_ce_runtime.py:283]: done in [7445df6](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/3/commits/7445df611c436a48b68610f086eb06f3e23b5c5d)

**Pull request merged**